lib/vtls/sectransp.c: Specify cipher name for Mac Secure Transport back-end

mkolechkin opened the pull request:
Add parser for CURLOPT_SSL_CIPHER_LIST option for Secure Transport (ST)
back-end. Similar to NNS and GSKit back-ends, new code parses string
value and configures ST library to use those ciphers for communication.

Create cipher spec data structure and initialize array of specs with
cipher number, names and flag. That array allows to add/modify ciphers
in one place instead of several functions in code. The flag is binary
value for backward compatibility with an existing wrapper
implementation. It can be extended later for other purposes like
support of specific TLS version.

Code change required some analysis of macOS SDK versions, we need to
know when one or another value of enum in macOS SDK was introduced,
and some other data. I created the TLS-CIPHERS.md file to summarize
that data I found and to add some more information about ciphers
could useful for others. That document is not completely ready yet
and I just want your opinion on it.

bagder commented:
Isn't the cipher names and info better put in `CIPHERS.md` where we have the name info for other TLS backends?

mkolechkin commented:
> Isn't the cipher names and info better put in `CIPHERS.md` where we have the name info for other TLS backends?
Yes, I just want to finish the table first and then merge it to the existing file. It is more convenient for me right now. The new table has more information and I prefer to send it as a new file. If reviewers agree the information is useful and format is good enough, etc, I'll definitely merge it into the CIPHERS.md

Question. I did not work on wide tables in .MD format before. Looks it is not very nice to scroll it right/left on the github. Do you see better way to present that information? Maybe split it to several tables or use some HTML tables instead?
I'll read more about tables in MD files. Maybe I can reduce font size or something like that.

Thank you for review! :)

nickzman commented on lib/vtls/sectransp.c:
```diff
@@ -138,6 +139,474 @@ struct ssl_backend_data {
+             CIPHER_WEAK_ANON_AUTH),
+  CIPHER_DEF(TLS_ECDH_anon_WITH_AES_256_CBC_SHA,                   /* 0xC019 */
+             CIPHER_WEAK_ANON_AUTH),
+#endif /* CURL_BUILD_MAC_10_6 || CURL_BUILD_IOS */
```
I don't see TLS_DH_anon_WITH_3DES_EDE_CBC_SHA on here. That weak cipher-suite needs to be disabled.

mkolechkin replied:
It is in lines 293-294, cipher number 0x1B.
Macro CIPHER_DEF_SSLTLS defines cipher with 2 names, one starts with "SSL_" prefix, another with "TLS_", and enum is referenced by SSL_ prefix. Last part we can skip from discussion for now.
Two names are done only for backward compatibility with previous/existing code.
Existing code has two functions `SSLCipherNameForNumber` and `TLSCipherNameForNumber` to return a cipher name by its number. First one is called when negotiated communication protocol is SSL. Second function is called for TLS communication. Result for cipher numbers 0x01 to 0x1B is different, depending on function. First function returns name starting with "SSL_" prefix while second one for the same numbers return name with prefix "TLS_". For example, name of the cipher number 0x01 is "SSL_RSA_WITH_NULL_MD5" when returned by the function `SSLCipherNameForNumber` and the same number mapped to "TLS_RSA_WITH_NULL_MD5" by the second function.
I wanted to keep it the same way and each cipher has 2 names: field `name` points to TLS_ name, field `ssl_name` points to SSL_ name. Those names are used for logging only.
Macro CIPHER_DEF_SSLTLS defines ciphers with both names but its argument is not the real number constant, it is a constant without TLS_ or SSL_ prefix. The macro concatenates prefixes to define both names, as well as concatenates parameter with SSL_ prefix to get the enum number properly. It is a C-style for some code generation, not the best style but I did not find better one.
You did not find `TLS_DH_anon_WITH_3DES_EDE_CBC_SHA` constant but you can find `DH_anon_WITH_3DES_EDE_CBC_SHA` in the new code.

I think better approach would be to simplify all that dance and drop names starting with "SSL_", always return "TLS_" name. I don't know how important backward compatibility for logging. I found that OpenSSL has some trace reporting code and they always print there "TLS_" names. If reviewers are fine with it, I'll simplify the code and make it always returning "TLS_" names, and all constants in the code will be the full constant like `TLS_DH_anon_WITH_3DES_EDE_CBC_SHA` not the `DH_anon_WITH_3DES_EDE_CBC_SHA`. To be more precise, it will be SSL_DH_anon_WITH_3DES_EDE_CBC_SHA, but we can discuss it next time.

Those complications related to macOS SDK constants and support were one of the reasons to create table I added to another file. I had other reasons also. I just want to explain my thoughts and reasons here.
I think it would be better to split this change on 2: one for source code only, another will have only the new table with cipher mappings. The table needs more work to be done.
I very appreciate any ideas and opinions.
Thank you! :)

nickzman commented on lib/vtls/sectransp.c:
```diff
@@ -138,6 +139,474 @@ struct ssl_backend_data {
+#define CIPHER_WEAK_IDEA_ENCRIPTION TRUE
+#define CIPHER_WEAK_ANON_AUTH       TRUE
+#define CIPHER_STRONG_ENOUGH        FALSE
+#define CIPHER_STRONG_ARGUABLY_3DES FALSE
```
I'd be in favor of disabling 3DES cipher-suites by default. 3DES is vulnerable to the Sweet32 attack.

nickzman commented on docs/TLS-CIPHERS.md:
```diff
@@ -0,0 +1,440 @@
+
+Draft notes to be finished later.
+Below I keep some notes to be finished later. Please ignore them for now.
+'MacOS SDK' - version of MacOS SDK when protocol number was 
```
That's "macOS", not "MacOS".

nickzman commented on docs/TLS-CIPHERS.md:
```diff
@@ -0,0 +1,440 @@
+column #13 - 'Notes' - some notes about cipher suite
+
+
+| Cipher | SSL 3.0 name                           | TLS IANA name                                 | OpenSSL cipher name           | wolfSSL cipher name            | GSKit cipher name | Network Security                      | MacOS    | TLS 1.0     | TLS 1.1     | TLS 1.2     | TLS 1.3    | Notes |
```
Is anyone else able to read this table? My Markdown reader (Typora) can't make any sense of it.

mkolechkin replied:
Probably table is too wide. I don't know if it can be solved unless broken down to several tables.
I looked at the result only in browser, hope the following link works https://github.com/curl/curl/pull/6464/files?short_path=1c41cba
Horizontal scrolling is very inconvenient. I am open to any suggestion.

nickzman reviewed (changes requested):
Thank you for your contribution. I've got a few changes I'd like you to make.

nickzman commented on lib/vtls/sectransp.c:
```diff
@@ -138,6 +139,474 @@ struct ssl_backend_data {
+     Marked in the code below as weak.
+ DES and IDEA encryption - rfc5469, has list of deprecated ciphers.
+     Marked in the code below as weak.
+ Anonymous Deffie-Hellman authentication and anonymous elliptic curve
```
It's "Diffie-Hellman."

nickzman commented on lib/vtls/sectransp.c:
```diff
@@ -1386,6 +1312,196 @@ set_ssl_version_min_max(struct connectdata *conn, int sockindex)
+    }
+    if(!cipher_found) {
+      /* It would be more human readable if we print the wrong cipher name
+         but we don't want to allocate an additional memory and copy the name
```
That should say "allocate any additional memory".

nickzman reviewed (changes requested):
Good; the latest changes are no longer generating a failing grade when I try them on howsmyssl.com's analyzer API. I found some typos in the code.

nickzman commented on lib/vtls/sectransp.c:
```diff
@@ -1818,122 +2132,12 @@ static CURLcode sectransp_connect_step1(struct Curl_easy *data,
-    failf(data, "SSL: SSLSetEnabledCiphers() failed: OSStatus %d", err);
-    return CURLE_SSL_CIPHER;
+  else {
+    sectransp_set_selected_ciphers(conn, sockindex, ciphers);
```
We need to check to see if this returns CURLE_SSL_CIPHER here, and if it's returned, set the default ciphers. Otherwise, there's a bug where passing in a nonsense or unsupported cipher-suite will enable ST's default ciphers, including the insecure ones.

mkolechkin replied:
Thank you, good catch.
I think it would be better not to switch to default ciphers but just fail.
I'll send version with your suggestion, but based on behavior of other back-ends (e.g. OpenSSL back-end), if specified cipher name is not valid, it just fails. That way it clearly tells to caller that parameter is wrong and doesn't mask the problem. I think we better do the same.
I'll send both variants, first your suggestion, second the one which fails.

nickzman commented on lib/vtls/sectransp.c:
```diff
@@ -1386,6 +1479,228 @@ set_ssl_version_min_max(struct Curl_easy *data, struct connectdata *conn,
+        ssl_name = TRUE;
+      }
+    }
+    /* Iterate through the ciphers table and look for the cipher */
```
Typo; that should be "cipher table".

nickzman commented on lib/vtls/sectransp.c:
```diff
@@ -1386,6 +1479,228 @@ set_ssl_version_min_max(struct Curl_easy *data, struct connectdata *conn,
+
+    /* In the future we can add support of cipher groups like cipher suites
+       supported by TLS1.0 or TLS1.3 etc. It would be subset of similar lists
+       or keywords supprted by OpenSSL. We do not need to implement complicated
```
Typo; that should be "keywords supported by OpenSSL".

nickzman commented on lib/vtls/sectransp.c:
```diff
@@ -1386,6 +1479,228 @@ set_ssl_version_min_max(struct Curl_easy *data, struct connectdata *conn,
+       protocol name (SSL_ or TLS_), if the 4th symbol is different then we
+       need to treat it as an alias name or a group name.
+       If it matches to a group name then we fill our array with all ciphers
+       in the group. This is a good place we can branch those two cases.
```
That should be "a good place where we can branch those two cases".

mkolechkin replied:
After some time and thought iterations, I reconsidered the point about cipher groups.
It requires more maintenance in the future of the source code and documentation, and doesn't give much advantage, from my point of view. I think I better remove that comment. If in the future we'll need to add that feature, we'll find place to add the new code.
The main purpose of that comment on my side was to discuss that feature or think more about it. I did the latter and made some decision.

Another point for me to think a bit more was to keep or remove SSL names from the code.
Existing code maps cipher number to the name only for logging. Also, first ciphers are mapped to 2 names: TLS_ name and SSL_ name. That is only for logging, and I looked at the OpenSSL code. They use only TLS names for logging (openssl-1.1.1h/ssl/t1_trce.c) and probably for some purposes, I did not find SSL_ names, only TLS_.
I think we also can drop SSL_ names. On the other hand, if we drop them, we break compatibility with existing version. I doubt someone relies on cipher names in logs for some automation or text matching, but I'd like to know your opinion.
I'll update PR with 2 versions, one addressing your comments, and another will remove SSL names on top of it.
Please take a look, when they'll be ready and let me know if you are ok to remove SSL names.

Thank you.

nickzman commented on lib/vtls/sectransp.c:
```diff
@@ -1386,6 +1479,228 @@ set_ssl_version_min_max(struct Curl_easy *data, struct connectdata *conn,
+      return !ciphertable[i].weak;
+    }
+  }
+  /* If cipher is not in our list, assume it is a new one
```
That should be "If the cipher is not in our list".

nickzman commented on lib/vtls/sectransp.c:
```diff
@@ -138,6 +139,641 @@ struct ssl_backend_data {
+     All of those filters were considered weak because they contain a weak
+     algorithm like DES, RC2 or RC4, and already considered weak by other
+     criteria.
+ 3DES - NIST deprecated it and going to retire it by 2023
```
That should be "and is going to retire it".

nickzman reviewed (changes requested):
You're very close; you just have typos and a security flaw to cover now. Again, thanks for your work on this so far.

mkolechkin commented:
I made several commits, probably I better summarize them here.

I updated PR based on comments, commit cc7f8f7 addresses most of them.
Then I thought a little bit and made changes: removed ssl_name and slightly modified error handling: fail on error instead of switch to default ciphers. Please take a look. If those changes are not reasonable, I can bring back ssl_name and default ciphers behavior on error. Commit b0fb321
The last commit cb84af2 was to fix iOS code guards. I looked at old SDKs (https://sskaje.me/ios-sdk-archive/) and found last cipher constants added to macOS SDK 10.15, they introduced to iOS SDK 13 (I have it on my machine and checked difference with v 11 available on the site I mentioned above). Also corrected iOS SDK guards for cipher constants as they were before my change

Please take a look.

nickzman commented on lib/vtls/sectransp.c:
```diff
@@ -295,586 +929,23 @@ static OSStatus SocketWrite(SSLConnectionRef connection,
-      break;
-#endif /* CURL_BUILD_MAC_10_13 || CURL_BUILD_IOS_11 */
+  /* The first ciphers in the ciphertable are continuos. Here we do small
+     optimization and instead of loop directly get SSL name by cipher number.
```
Typos; that should say "The first ciphers in the table are continuous. Here, we do a small optimization, and instead of loop directly, we get the TLS name by cipher number."

mkolechkin replied:
I think I better change the wording a bit. Optimization is really small, I just don't know the CURL coding style well enough. I initially wanted to make a note in 2 places: the first place is the table definition asking to keep the sequence, and the second place is here.
I made it a bit different, please take a look.
Thank you

nickzman reviewed (changes requested):
I made one more pass through this, and found one more typo in the comments. Otherwise, this is looking good.

nickzman reviewed (approved):
All right, this looks good. Can you squash this please? @bagder , do you want to review?

mkolechkin commented:
> All right, this looks good. Can you squash this please?
Thank you for your help!
I hope I squashed it right, did it the first time.

mkolechkin commented:
@bagder @nickzman 
Hi Daniel and Nick,
Do I need to do something more for this change to be merged into the master branch?
Thank you
Mike

bagder commented:
I'm ready to merge once the feature window opens again. If all goes well, that's on Monday April 5th.

bagder commented:
Thanks!

mkolechkin commented:
Thank you, Daniel and Nick!
